File: riveredge-backend/src/apps/master_data/models/drawing_loan.py
"""图档借阅单"""

from tortoise import fields

from core.models.base import BaseModel


class DrawingLoan(BaseModel):
    class Meta:
        table = "apps_master_data_drawing_loans"
        table_description = "基础数据管理 - 图档借阅单"
        indexes = [
            ("tenant_id",),
            ("code",),
            ("status",),
            ("created_at",),
        ]

    id = fields.IntField(pk=True, description="主键ID")
    code = fields.CharField(max_length=50, description="借阅单号")
    name = fields.CharField(max_length=200, description="借阅单名称")
    purpose = fields.CharField(max_length=500, null=True, description="借阅用途")
    due_at = fields.DatetimeField(description="应还时间")
    status = fields.CharField(
        max_length=20,
        default="Draft",
        description="Draft/Pending/Borrowed/Returned",
    )
    returned_at = fields.DatetimeField(null=True, description="归还时间")
    returned_by = fields.IntField(null=True, description="归还人ID")
    returned_by_name = fields.CharField(max_length=100, null=True, description="归还人姓名")
    deleted_at = fields.DatetimeField(null=True, description="软删除")
    created_by = fields.IntField(null=True, description="创建人ID")
    created_by_name = fields.CharField(max_length=100, null=True, description="创建人姓名")
    updated_by = fields.IntField(null=True, description="更新人ID")
    updated_by_name = fields.CharField(max_length=100, null=True, description="更新人姓名")

    def __str__(self) -> str:
        return self.code


class DrawingLoanLine(BaseModel):
    class Meta:
        table = "apps_master_data_drawing_loan_lines"
        table_description = "基础数据管理 - 图档借阅明细"
        indexes = [
            ("tenant_id",),
            ("loan_id",),
            ("drawing_id",),
        ]

    id = fields.IntField(pk=True, description="主键ID")
    loan_id = fields.IntField(description="借阅单ID")
    drawing_id = fields.IntField(description="图纸ID")
    drawing_uuid = fields.CharField(max_length=36, description="图纸UUID")
    drawing_code = fields.CharField(max_length=50, description="图号")
    drawing_name = fields.CharField(max_length=200, description="图纸名称")
    drawing_revision = fields.CharField(max_length=20, description="修订版")
    security_level = fields.CharField(max_length=20, description="借出时密级")

File: riveredge-frontend/src/apps/master-data/services/drawingLoan.ts
/**
 * 图档借阅单与密级授权 API
 */

import { api } from '../../../services/api';

export type DrawingLoanStatus = 'Draft' | 'Pending' | 'Borrowed' | 'Returned';
export type DrawingSecurityLevel = 'public' | 'internal' | 'secret' | 'confidential';

export interface DrawingLoanLine {
  id: number;
  drawingId: number;
  drawingUuid: string;
  drawingCode: string;
  drawingName: string;
  drawingRevision: string;
  securityLevel: DrawingSecurityLevel;
}

export interface DrawingLoan {
  id: number;
  uuid: string;
  tenantId: number;
  code: string;
  name: string;
  purpose?: string;
  dueAt: string;
  status: DrawingLoanStatus;
  returnedAt?: string;
  returnedByName?: string;
  createdByName?: string;
  updatedByName?: string;
  createdAt: string;
  updatedAt: string;
  lines: DrawingLoanLine[];
}

export interface DrawingLoanListResponse {
  data: DrawingLoan[];
  total: number;
}

export interface DrawingClearance {
  userId: number;
  userName: string;
  securityLevel: DrawingSecurityLevel;
  updatedByName?: string;
  updatedAt: string;
}

export interface DrawingClearanceListResponse {
  data: DrawingClearance[];
  total: number;
}

const BASE = '/apps/master-data/process/drawing-loans';
const CLEARANCE_BASE = '/apps/master-data/process/drawing-clearances';

export const drawingLoanApi = {
  list: async (params?: {
    skip?: number;
    limit?: number;
    status?: string;
    keyword?: string;
  }): Promise<DrawingLoanListResponse> => {
    return api.get<DrawingLoanListResponse>(BASE, { params });
  },

  get: async (uuid: string): Promise<DrawingLoan> => {
    return api.get<DrawingLoan>(`${BASE}/${uuid}`);
  },

  create: async (data: {
    code?: string;
    name: string;
    purpose?: string;
    dueAt: string;
    lines: Array<{ drawingUuid: string }>;
  }): Promise<DrawingLoan> => {
    return api.post<DrawingLoan>(BASE, data);
  },

  update: async (
    uuid: string,
    data: { name?: string; purpose?: string; dueAt?: string; lines?: Array<{ drawingUuid: string }> },
  ): Promise<DrawingLoan> => {
    return api.put<DrawingLoan>(`${BASE}/${uuid}`, data);
  },

  delete: async (uuid: string): Promise<void> => {
    return api.delete(`${BASE}/${uuid}`);
  },

  submit: async (uuid: string): Promise<DrawingLoan> => {
    return api.post<DrawingLoan>(`${BASE}/${uuid}/submit`, {});
  },

  approve: async (uuid: string): Promise<DrawingLoan> => {
    return api.post<DrawingLoan>(`${BASE}/${uuid}/approve`, {});
  },

  reject: async (uuid: string): Promise<DrawingLoan> => {
    return api.post<DrawingLoan>(`${BASE}/${uuid}/reject`, {});
  },

  revoke: async (uuid: string): Promise<DrawingLoan> => {
    return api.post<DrawingLoan>(`${BASE}/${uuid}/revoke`, {});
  },

  complete: async (uuid: string): Promise<DrawingLoan> => {
    return api.post<DrawingLoan>(`${BASE}/${uuid}/complete`, {});
  },

  listClearances: async (): Promise<DrawingClearanceListResponse> => {
    return api.get<DrawingClearanceListResponse>(CLEARANCE_BASE);
  },

  upsertClearance: async (data: {
    userId: number;
    securityLevel: DrawingSecurityLevel;
  }): Promise<DrawingClearance> => {
    return api.put<DrawingClearance>(CLEARANCE_BASE, data);
  },

  deleteClearance: async (userId: number): Promise<void> => {
    return api.delete(`${CLEARANCE_BASE}/${userId}`);
  },
};
